src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -883,12 +883,21 @@ private BoundExpression CreateAndPopulateList(BoundCollectionExpression node, Ty
                 }
             }
 
+            // Create a temp for the knownLength
+            BoundAssignmentOperator assignmentToTemp;
+            BoundLocal? knownLengthTemp = null;
+
             BoundObjectCreationExpression rewrittenReceiver;
-            if (useKnownLength && elements.Length > 0 && !useOptimizations)
+            if (useKnownLength && elements.Length > 0)
             {
-                // List<ElementType> list = new(N + s1.Length + ...);
+                // int knownLengthTemp = N + s1.Length + ...;
+                knownLengthTemp = _factory.StoreToTemp(GetKnownLengthExpression(elements, numberIncludingLastSpread, localsBuilder), out assignmentToTemp, isKnownToReferToTempIfReferenceType: true);
+                localsBuilder.Add(knownLengthTemp);
+                sideEffects.Add(assignmentToTemp);
+
+                // List<ElementType> list = new(knownLengthTemp);
                 var constructor = ((MethodSymbol)_factory.WellKnownMember(WellKnownMember.System_Collections_Generic_List_T__ctorInt32)).AsMember(collectionType);
-                rewrittenReceiver = _factory.New(constructor, ImmutableArray.Create(GetKnownLengthExpression(elements, numberIncludingLastSpread, localsBuilder)));
+                rewrittenReceiver = _factory.New(constructor, ImmutableArray.Create<BoundExpression>(knownLengthTemp));
```

Comment:
Let's make sure to verify that the temp is not present in the emitted IL in the case that 'useOptimizations' is false in this method. (i.e. the subsequent stack optimizer pass should be able to eliminate it when compiling in release mode.)